### DISCUSSION TITLE
Inquiry Regarding grantToken Method Implementation

### DISCUSSION THREAD

**Discussion by JhaSwechha** [keycloak/keycloak#27754] [Q&A] [answered]
Hi Team,

We use a fork of your repo for our org and I'm working on a change in the token endpoint.  I'm referencing the API documentation for the `/realms/{realm}/protocol/openid-connect/token` endpoint and went through the codebase as well.

The code mentions an `AccessTokenResponse grantToken(@PathParam("realm") String realm, MultivaluedMap<String, String> map);` method in TokenService.java for the above api, but I haven't been able to locate its implementation.

Could you please point me towards the code that implements this method? 

I have spent a good amount of time trying to find the code where the implementation of **grantToken** is written as mentioned in the TokenService.java file but unable to find it.

It would be great help.

**Comment by JhaSwechha** [comment]
Can someone please help?

**Reply by danielFesenmeyer** [reply, marked as answer]
The implementation is in TokenEndpoint: https://github.com/keycloak/keycloak/blob/1e48cce3ae80d555bb3434b835dfd52d72658b16/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java#L111

**Comment by JhaSwechha** [comment]
Really kind of stuck here

**Reply by JhaSwechha** [reply]
Thank you!
